### PR TITLE
Add supported backward_dtype to OpInfo

### LINKED ISSUE #55601: Verify complex autograd is actually not supported in test_supported_backward
Today this test is skipped for complex dtypes if the op says it doesn't support complex backward:

https://github.com/pytorch/pytorch/blob/17e5ba44f1f67eb734fdf007c52c4374abbd71ab/test/test_ops.py#L70

But we should stop trusting ops and assert that attempting the operation throws an error.

### PR DESCRIPTION
**Pull request by walterddr**
Related to #55601.

- [x] removed complex autograd checker in `test_supported_backward`
- [x] created `backward_dtype[If<Device>]` that inherits from normal `dtype[If<Device>]` by default
- [x] removed all skip for backward test, instead added backward dtype
- [x] change complex autograd to a function call: `support_complex_autograd(device_type)` that depends on `backward_dtype*` since they essentially mean the same thing for complex types

TODO for next PR
- add `test_unsupported_backward` to verify they are actually unsupported.

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -4854,6 +4871,8 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    dtypes=all_types_and_complex_and(torch.bool, torch.bfloat16),
                    dtypesIfCPU=all_types_and_complex_and(torch.bool, torch.bfloat16),
                    dtypesIfCUDA=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
+                   backward_dtypesIfCPU=all_types_and(torch.bool, torch.bfloat16),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:4874]: This is interesting. How do you suggest we think of the backward dtypes vs. the "supports_complex_autograd" metadata? Would you recommend removing the latter? 

cc @anjali411

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -297,6 +310,13 @@ def supported_dtypes(self, device_type):
         else:
             return self.dtypes
 
+    def supported_backward_dtypes(self, device_type):
+        if device_type == 'cpu':
+            return self.backward_dtypesIfCPU
+        if device_type == 'cuda':
+            return self.backward_dtypesIfROCM if TEST_WITH_ROCM else self.backward_dtypesIfCUDA
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:317]: Would you also fix line 309 above while we're here?

**Review by mruberry — COMMENTED** [review]
Hey @walterddr! This looks great, with really good attention to detail. I'd like to hear your thoughts on supported backward dtypes vs the current supports_complex_autograd metadata. Maybe there's an opportunity to replace the attribute?

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (walterddr) [reply torch/testing/_internal/common_methods_invocations.py:317]: didn't even realize --> will try and if it incurs more test failure I will open a separate PR

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (walterddr) [reply torch/testing/_internal/common_methods_invocations.py:4874]: `supports_complex_autograd` is not only used on `test_supported_backward` but also used in other test in test_ops.py such as test_variant_consistency_eager, test_variant_consistency_jit.

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:4874]: sure, but that seems solvable

we could even add op.supports_autograd(device, dtype) -> bool as a function and replace supports_complex_autograd uses with it, I bet

I just wonder if there should be a "single source of truth" for whether an op supports complex autograd, or maybe there's an advantage to having both?

Comment (walterddr) [reply torch/testing/_internal/common_methods_invocations.py:4874]: q: is there a chance when an `op` supports complex_autograd, but not backward?

I think this is equivalent to ask if this claim is correct: "An (`op` x `dtype`) combo to support backward **if and only if** `op` support autograd for that `dtype`. if so we can treat them as equal and replace all support_complex_autograd with backward_dtype

Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py:4874]: > q: is there a chance when an `op` supports complex_autograd, but not backward?

I don't understand what you mean. afaik both those things mean the same thing

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:4874]: I like the idea of going exclusively to whether the op supports a dtype in backward vs. complex autograd, I'd also like to make a helper function, like `op.supports_autograd(device, dtype)`, to replace the existing queries to supports_complex_autograd

@anjali411, what would you think of that approach? It's more general, but it also loses the readable `supports_complex_autograd` field. Maybe we could add comments above supported backward dtypes describing complex autograd support to make it more readable?

Comment (walterddr) [reply torch/testing/_internal/common_methods_invocations.py:4874]: +1 on the helper function approach

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (walterddr) [reply torch/testing/_internal/common_methods_invocations.py:317]: extra 160+ test failure were found after making that move (tested in #56646). will create a separate PR

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:317]: Thanks for trying. I imagine there's a failure with how we're acquiring the ROCm dtypes. I think today, for example, we populate them from dtypes, for example, and I think some OpInfos are too aggressive in specifying their base dtypes. Maybe we should be acquiring the default ROCm dtypes not from the base "dtypes" but from the CUDA dtypes?

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (imaginary-person) [reply torch/testing/_internal/common_methods_invocations.py:4874]: For `pow`, backward is not supported for `float16` on CPU. It looks like this PR might help in such cases, and would also make it easier to document the dtypes an op supports autograd for.

Comment (walterddr) [reply torch/testing/_internal/common_methods_invocations.py:4874]: I've added a follow up TODO to the description. it seems like a pretty sizable change so will follow up with another PR.

**Comment by walterddr** [comment]
rocm failures are legit. figuring out a way to fix / finding a way to get the hardware env

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -199,6 +203,15 @@ def __init__(self,
...
         self.dtypesIfCUDA = set(dtypesIfCUDA) if dtypesIfCUDA is not None else self.dtypes
         self.dtypesIfROCM = set(dtypesIfROCM) if dtypesIfROCM is not None else self.dtypes
+
+        self.backward_dtypes = set(backward_dtypes) if backward_dtypes is not None else self.dtypes
+        self.backward_dtypesIfCPU = set(backward_dtypesIfCPU) if backward_dtypesIfCPU is not None else (
+            self.dtypesIfCPU if dtypesIfCPU is not None else self.backward_dtypes)
+        self.backward_dtypesIfCUDA = set(backward_dtypesIfCUDA) if backward_dtypesIfCUDA is not None else (
+            self.dtypesIfCUDA if dtypesIfCUDA is not None else self.backward_dtypes)
+        self.backward_dtypesIfROCM = set(backward_dtypesIfROCM) if backward_dtypesIfROCM is not None else (
+            self.dtypesIfROCM if dtypesIfROCM is not None else self.backward_dtypes)
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:213]: I think this needs an additional fallback layer to CUDA to resolve the ROCm issues

**Comment by walterddr** [comment]
d37d636 depends on #56646.

**Comment by mruberry** [comment]
@walterddr Is this ready for another review?

**Comment by walterddr** [comment]
could you please help take a look at #56646 first? It would be better if we get the fallback logic clear between the device types

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py:5291]: I agree with Mike that it would reduce the readability a little to remove `supports_complex_autograd` kwarg, but at the same time it's also bad to have two kwargs sending the same signal. I think it's ok to get rid of the `supports_complex_autograd`, if we create an issue where the forward is supported for complex, but the backward is not and link it in a comment wherever complex autograd is skipped for only backward.

**Comment by walterddr** [comment]
#56646 has been merged and just rebased. This should be ready for another review, thanks @mruberry !

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:5291]: @walterddr This PR looks very good, what would you think of extending it to eliminate `supports_complex_autograd` and file the follow-up issues @anjali411 mentioned?

I realize that might be a bit of a pain, but I think it would make the PR more complete. Also there are only six opinfos that current use supports_complex_autograd, which shouldn't be too bad.

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -305,6 +313,24 @@ def supported_dtypes(self, device_type):
...
             return self.dtypes
 
+    def supported_backward_dtypes(self, device_type):
+        if device_type == 'cpu':
+            return self.backward_dtypesIfCPU
+        if device_type == 'cuda':
+            return self.backward_dtypesIfROCM if TEST_WITH_ROCM else self.backward_dtypesIfCUDA
+        else:
+            return self.backward_dtypes
+
```
Comment (walterddr) [inline torch/testing/_internal/common_methods_invocations.py:323]:
```suggestion

```

File: test/test_ops.py [hunk 5]
```diff
@@ -362,7 +360,7 @@ def _test_inplace_preserve_storage(samples, variants):
     # TODO WARNING: inplace x {traced, scripted} not currently tested
     @_variant_ops(op_db)
     def test_variant_consistency_jit(self, device, dtype, op):
-        _requires_grad = op.supports_autograd and (dtype.is_floating_point or op.supports_complex_autograd)
+        _requires_grad = op.supports_autograd and (dtype.is_floating_point or op.supports_complex_autograd(torch.device(device).type))
```
Comment (walterddr) [inline test/test_ops.py:363]:
```suggestion
        _requires_grad = op.supports_autograd and (dtype.is_floating_point or
                                                   op.supports_complex_autograd(torch.device(device).type))
```

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -305,6 +313,23 @@ def supported_dtypes(self, device_type):
...
             return self.dtypes
 
+    def supported_backward_dtypes(self, device_type):
+        if device_type == 'cpu':
+            return self.backward_dtypesIfCPU
+        if device_type == 'cuda':
+            return self.backward_dtypesIfROCM if TEST_WITH_ROCM else self.backward_dtypesIfCUDA
+        else:
+            return self.backward_dtypes
+    def supports_complex_autograd(self, device_type):
```
Comment (walterddr) [inline torch/testing/_internal/common_methods_invocations.py:323]:
```suggestion

    def supports_complex_autograd(self, device_type):
```

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (walterddr) [reply torch/testing/_internal/common_methods_invocations.py:5292]: just made another change to convert `supports_complex_autograd(device_type)` as a function depends on backward dtypes. could you please kindly take another look @mruberry

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -5271,9 +5289,10 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    dtypes=all_types_and_complex_and(torch.bool, torch.bfloat16),
                    dtypesIfCPU=all_types_and_complex_and(torch.bool, torch.bfloat16),
                    dtypesIfCUDA=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
+                   backward_dtypesIfCPU=all_types_and(torch.bool, torch.bfloat16),
+                   backward_dtypesIfCUDA=all_types_and(torch.bool, torch.bfloat16),
                    safe_casts_outputs=True,
-                   assert_autodiffed=True,
-                   supports_complex_autograd=False),  # Reference: https://github.com/pytorch/pytorch/issues/48552
+                   assert_autodiffed=True),  # Reference: https://github.com/pytorch/pytorch/issues/48552
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5295]: Just move this comment above the "backward_dtypes" lines

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -3979,10 +4004,10 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            op=torch.linalg.det,
            aliases=('det', ),
            dtypes=floating_and_complex_types(),
+           backward_dtypes=floating_types(),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:4007]: Add a comment (ideally with a link to an issue) above these "backward_dtypes" lines where supports_complex_autograd was removed, just to keep the lack of complex autograd support visible

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -4579,10 +4599,10 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
     OpInfo('std',
            dtypes=floating_types_and(),
            dtypesIfCUDA=floating_and_complex_types_and(torch.half, torch.bfloat16),
+           backward_dtypesIfCUDA=floating_types_and(torch.half, torch.bfloat16),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:4602]: Comment here

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -5179,10 +5198,10 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
     OpInfo('var',
            dtypes=floating_types_and(),
            dtypesIfCUDA=floating_and_complex_types_and(torch.half, torch.bfloat16),
+           backward_dtypesIfCUDA=floating_types_and(torch.half, torch.bfloat16),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5201]: Comment here

File: torch/testing/_internal/common_methods_invocations.py [hunk 11]
```diff
@@ -5198,7 +5217,6 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
     OpInfo('logsumexp',
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5217]: Interesting, so this does support complex autograd now? That's awesome

**Review by mruberry — APPROVED** [review]
Really cool work, @walterddr! I think there may be an opportunity to extend on this, too, by adding a "test_unsupported_backward" in test_ops.py.

I made just a few small notes about adding a couple comments or moving existing comments slightly.

File: torch/testing/_internal/common_methods_invocations.py (hunk 11 above)
Comment (walterddr) [reply torch/testing/_internal/common_methods_invocations.py:5217]: well I dont see why that flag was needed in the first place, b/c no `dtypes*` includes complex type.
